### PR TITLE
Make BART more ONNX friendly

### PR DESCRIPTION
**Pull request by mfuntowicz**
**Major concerns:** 
1. ONNX complains about Python raw integer usage.
2. ONNX doesn't support boolean indexing with something else than vector. The code was using 2D Tensor indices (batch, token). 

**PR workarounds:**
1. Remove the call to `len(..)` and prefer the use of `.size(-1)`
2. Attempt to index the output tensors to retrieve only the **last** EOS token over the sequence axis through `.index_select(..)`

### CODE REVIEW

File: src/transformers/modeling_bart.py [hunk 1]
```diff
@@ -1199,10 +1199,13 @@ def forward(
...
-        if len(torch.unique(eos_mask.sum(1))) > 1:
+        if torch.unique(eos_mask.sum(1)).size(-1) > 1:
             raise ValueError("All examples must have the same number of <eos> tokens.")
-        sentence_representation = x[eos_mask, :].view(x.size(0), -1, x.size(-1))[:, -1, :]
-        logits = self.classification_head(sentence_representation)
+
+        # Attempt to gather the latest eos_token representation in an ONNX compatible way
+        # (-1: remove the batch indexes from the vector, -1: Take the last occurrence of eos_token)
+        sentence_representation = torch.index_select(x, 1, eos_mask.nonzero(as_tuple=False).t()[-1, -1])
+        logits = self.classification_head(sentence_representation).squeeze(1)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_bart.py:1208]: do we have to do the `.squeeze(1)` here? Is this fully backwards compatible?

**Comment by LysandreJik** [comment]
I believe @mfuntowicz ran the slow tests on this one and there were no failures, but this PR is so cryptic I don't understand what's happening.
